### PR TITLE
Fix RPC address clashes on local multi-node testnet

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem
Fullnodes created in a local testnet figure out what RPC address to use from the json config file which has a static address for RPC. This results in a panic when running two or more nodes in a local testnet since the client's RPC requests can be served by the wrong node.

#### Summary of Changes
Make fullnodes search for available rpc ports instead of using the one from the json config. Clients already find nodes via gossip anyway. 

Fixes #

### CODE REVIEW

File: src/bin/fullnode.rs [hunk 1]
```diff
@@ -116,7 +110,27 @@ fn main() {
         }
         Some(port_number)
     } else {
-        None
+        match find_available_port_in_range((8899, 8999)) {
```
Comment (rob-solana) [inline src/bin/fullnode.rs]: magic numbers...

**Comment by rob-solana** [comment]
how about some tests?  something that fails without your change.

File: src/bin/fullnode.rs (hunk 1 above)
Comment (sagar-solana) [reply src/bin/fullnode.rs]: Forgot to comment on this. I wanted suggestions. I can make it the full range for now if you think thats okay? There's a #define for the full range of ports.

**Comment by sagar-solana** [comment]
I can try to come up with some tests, but it is a race condition after all...

File: src/bin/fullnode.rs (hunk 1 above)
Comment (rob-solana) [reply src/bin/fullnode.rs]: prolly want to start commenting on magics where we define them instead of where we use 'em

**Comment by rob-solana** [comment]
I don't think it's a race, it's a "feature" of the library we're using to re-use addresses if there are multiple serving threads.

Also: please make sure you take UDP or TCP into account for the find_available_port... stuff.  most code uses UDP and the rpc stuff is TCP

File: src/bin/fullnode.rs (hunk 1 above)
Comment (sagar-solana) [reply src/bin/fullnode.rs]: You're right. I meant comment on here asking for feedback :)

**Comment by sagar-solana** [comment]
Okay I see what you mean. I think there's two levels to this problem. What would happen if reuse address wasn't set on the rpc library? would the bind fail and the rpc service be lost on that node (problem) or would it only allow the latest server to bind to it (also a problem)? This PR tries to avoid clashes altogether. To write a test for this would be to test the side effect of both nodes using the same rpc address and that test would be checking for a race condition of sorts. Am I understanding this wrong? 

I lifted the function from somewhere, I'll take a look at the socket types as well. Thanks!

**Comment by CriesofCarrots** [comment]
This looks reasonable to me. `find_available_port_in_range` is for TCP, so you should be okay there. However I noticed an issue in that method that could prevent it finding open ports, particularly in the case of small ranges. I'll open a PR with a fix.

**Pull request merged**